community/planning/service_timer.md

```diff
@@ -0,0 +1,244 @@
+
+```rust
+pub struct MyTimerFilter { ... }
+
+impl TimerFilter for MyTimerFilter {
+    fn filter(&self) -> Result<Vec<FullyQualifiedServiceId>, InternalError> {
+         // query the database and return a list of service ids which should be woken up
+    }
+}
+```
+
+### Runtime execution and integration within Splinter
+
+A `Timer` handles execution of both `TimerFilter`s and `TimerHander`s. The
+`Timer` wakes up periodically and runs a `TimerFilter` for each service type
+supported (example service types: echo, scabbardv2). Each filter returns a list
+of service IDs and for each service ID returned, the `Timer` executes a
+`TimerHandler` of the corresponding type.
+
+## Reference-level Explanation
+
+### TimerFilter Trait
+The `TimerFilter` is used to determine the list of `FullyQualifiedServiceId`
+that needs to be handled. Each service type will need their own `TimerFilter`.
+
+```rust
+pub trait TimerFilter: Routable{
+    fn filter(&self) -> Result<Vec<FullyQualifiedServiceId>, InternalError>;
+}
+```
+
+The `TimerFilter` must be `Routable` meaning it has an associated service types.
+
+```rust
+pub trait Routable {
+    fn service_types(&self) -> &[ServiceType];
+}
+```
+
+For example, the `EchoService` will return the list of all services that have a
+peer and are in the `Finalized` state. See the
+[EchoService]({% link community/planning/echo_service_design.md%}).
+for more information
+
+### TimerHandler Trait
+The `TimeHandler` is in charge of executing any work that should be done on some
+interval.
+
+```rust
+pub trait TimerHandler {
+    type Message;
+
+    fn handle_timer(
+        &mut self,
+        sender: &dyn MessageSender<Self::Message>,
+        service: FullyQualifiedServiceId,
+    ) -> Result<(), InternalError>;
+
+    fn into_handler<C, R>(
+self,
+converter: C
+) -> IntoTimerHandler<Self, C, Self::Message, R>
+    where
+        Self: Sized,
+        C: MessageConverter<Self::Message, R>,
+    {
+        IntoTimerHandler::new(self, converter)
+    }
+}
+```
+
+The use of  `into_handler` is described in **Link to avoid bytes doc to explain
+into_handler**
+
+### TimerHandlerFactory Trait
+`TimerHandlerFactory` will be used to create new handlers in the `Timer` so they
+can be passed to a threadpool for execution. The handle must be Send and be
+cloneable through the use of `clone_box`
+
+```rust
+pub trait TimerHandlerFactory: Send {
+    type Message;
+
+    fn new_handler(
+&self
+) -> Result<Box<dyn TimerHandler<Message = Self::Message>>, InternalError>;
+
+    fn clone_box(&self) -> Box<dyn TimerHandlerFactory<Message = Self::Message>>;
+}
+```
+
+## Timer Struct
+The `Timer` is in charge of checking the configured `TimerFilters` for pending
+work.
+
+![Timer Struct Diagram]({% link images/timer_struct.svg %}
+"Timer"){:.centered}
+
+```rust
+impl Timer {
+    pub fn new(
+        filters: Vec<(
+    Box<dyn TimerFilter + Send>,
+    Box<dyn TimerHandlerFactory<Message = Vec<u8>>>,
+)>;
```

Comment:
should this be a comma instead of a semicolon?

Reply:
fixed